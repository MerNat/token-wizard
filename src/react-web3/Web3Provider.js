import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Web3 from 'web3'
import { Loader } from '../components/Common/Loader'
import logdown from 'logdown'

const ONE_SECOND = 1000
const logger = logdown('TW:Web3Provider')

class Web3Provider extends Component {
  constructor(props, context) {
    super(props, context)

    this.state = {
      selectedAccount: null,
      web3: null,
      render: false
    }

    this.web3 = null
    this.interval = null
  }

  render() {
    const { web3UnavailableScreen: Web3UnavailableScreen } = this.props

    let renderContainer = false

    if (this.state.render) {
      if (this.state.web3) {
        renderContainer = this.props.children
      } else {
        // eslint-disable-next-line
        renderContainer = <Web3UnavailableScreen/>
      }
    } else {
      // eslint-disable-next-line
      renderContainer = <Loader show={true}></Loader>
    }
    // eslint-disable-next-line
    return ( renderContainer )
  }

  componentDidMount() {
    this.checkWeb3()
    this.initPoll()

    // Wait 1 second to render
    setTimeout(() => {
      this.setState({ render: true })
    }, 500)
  }

  checkWeb3() {
    const setWeb3 = () => {
      try {
        window.web3 = new Web3(window.web3.currentProvider)
        this.setState({
          web3: window.web3
        })
        this.fetchAccounts()
      } catch (err) {
        logger.log('There was a problem fetching accounts', err)
      }
    }
    window.addEventListener('load', () => {
      const { ethereum } = window
      // Modern dapp browsers...
      if (ethereum) {
        window.web3 = new Web3(ethereum)
        try {
          // Request account access if needed
          ethereum.enable().then(() => {
            if (!this.state.web3) {
              setWeb3()
            }
          })
        } catch (error) {
          // User denied account access...
          logger.log('User denied account access')
        }
      }
      // Legacy dapp browsers...
      else if (window.web3) {
        setWeb3()
      }
      // Non-dapp browsers...
      else {
        logger.log('Non-Ethereum browser detected. You should consider trying a wallet!')
      }
    })
  }

  componentWillUnmount() {
    clearInterval(this.interval)
  }

  initPoll = () => {
    if (!this.interval) {
      this.interval = setInterval(this.fetchAccounts, ONE_SECOND)
    }
  }

  fetchAccounts = () => {
    const { web3 } = this.state
    const { onChangeAccount } = this.props

    if (!web3 || !web3.eth) {
      return
    }

    return web3.eth.getAccounts().then(accounts => {
      if (!accounts || !accounts.length) {
        return
      }

      let curr = this.state.selectedAccount
      let next = accounts[0]
      curr = curr && curr.toLowerCase()
      next = next && next.toLowerCase()

      const didChange = curr && next && curr !== next

      if (didChange && typeof onChangeAccount === 'function') {
        onChangeAccount(next)
      }

      this.setState({
        selectedAccount: next || null
      })
    })
  }

  getChildContext() {
    return {
      web3: this.state.web3,
      selectedAccount: this.state.selectedAccount
    }
  }
}

Web3Provider.childContextTypes = {
  web3: PropTypes.object,
  selectedAccount: PropTypes.string
}

export default Web3Provider
