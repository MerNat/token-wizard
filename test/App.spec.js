import React from 'react'
import Adapter from 'enzyme-adapter-react-15'
import { configure, mount, render } from 'enzyme'
import { MemoryRouter } from 'react-router'
import renderer from 'react-test-renderer'
import { Web3Provider } from '../src/react-web3'
import { deploymentStore } from '../src/stores'
import App from '../src/App'

configure({ adapter: new Adapter() })

describe('App Index', () => {
  const stores = { deploymentStore }

  beforeAll(() => {
    jest.useFakeTimers()
  })

  afterEach(() => {
    jest.clearAllTimers()
  })

  it(`should render app screen`, () => {
    // Given
    const component = renderer.create(
      <MemoryRouter initialEntries={['/']}>
        <App {...stores} />
      </MemoryRouter>
    )

    // When
    const tree = component.toJSON()

    // Then
    expect(tree).toMatchSnapshot()
  })

  it('renders without crashing', () => {
    // Given
    const wrapper = mount(
      <MemoryRouter initialEntries={['/']}>
        <App {...stores} />
      </MemoryRouter>
    )

    // Given
    const tree = wrapper.html()

    // Then
    expect(tree).toMatchSnapshot()
  })

  it('should render screen with render without throwing an error', () => {
    // When
    const wrapper = render(
      <MemoryRouter initialEntries={['/']}>
        <App {...stores} />
      </MemoryRouter>
    )

    // Given
    const tree = wrapper.html()

    // Then
    expect(tree).toMatchSnapshot()
  })

  it(`should not find web3Provider at home`, () => {
    // Given
    const wrapper = mount(<App {...stores} />)
    jest.runTimersToTime(2000)

    // When
    const web3ProviderWrapper = wrapper.find(Web3Provider)

    // Then
    expect(web3ProviderWrapper).toHaveLength(0)
  })

})
