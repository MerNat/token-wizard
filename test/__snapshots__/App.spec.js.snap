// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Index renders without crashing 1`] = `"<div><div><section class=\\"hm-Home\\"><div class=\\"hm-Home_Scroll\\"><div class=\\"hm-Home_Top\\"><a class=\\"sw-LogoPrimary\\" href=\\"/\\"></a></div><div class=\\"hm-Home_MainInfoContainer\\"><div class=\\"hm-Home_MainInfo\\"><div class=\\"hm-Home_MainInfoTextContainer\\"><h1 class=\\"hm-Home_MainInfoTitle\\">Welcome to Token Wizard</h1><p class=\\"hm-Home_MainInfoDescription\\">Token Wizard is a client side tool to create ERC20 token and crowdsale in five steps. It helps you to publish contracts on the Ethereum network, create a crowdsale page with stats. For participants, the wizard creates a page to contribute into the campaign.</p><p class=\\"hm-Home_MainInfoPoweredBy\\"><!-- react-text: 13 -->Token Wizard is powered by <!-- /react-text --><a href=\\"https://github.com/auth-os/beta\\">Auth-os</a><!-- react-text: 15 -->.<!-- /react-text --></p></div><div class=\\"hm-Home_MainInfoButtonContainer\\"><button class=\\"hm-Home_BtnNew\\">New crowdsale</button><div class=\\"hm-Home_BtnChoose\\">Choose Crowdsale</div></div></div></div></div></section></div><div class=\\"css-nm2dr3\\"><span></span></div></div>"`;

exports[`App Index should render app screen 1`] = `
<div>
  <div>
    <section
      className="hm-Home"
    >
      <div
        className="hm-Home_Scroll"
      >
        <div
          className="hm-Home_Top"
        >
          <a
            className="sw-LogoPrimary"
            href="/"
            onClick={[Function]}
          />
        </div>
        <div
          className="hm-Home_MainInfoContainer"
        >
          <div
            className="hm-Home_MainInfo"
          >
            <div
              className="hm-Home_MainInfoTextContainer"
            >
              <h1
                className="hm-Home_MainInfoTitle"
              >
                Welcome to Token Wizard
              </h1>
              <p
                className="hm-Home_MainInfoDescription"
              >
                Token Wizard is a client side tool to create ERC20 token and crowdsale in five steps. It helps you to publish contracts on the Ethereum network, create a crowdsale page with stats. For participants, the wizard creates a page to contribute into the campaign.
              </p>
              <p
                className="hm-Home_MainInfoPoweredBy"
              >
                Token Wizard is powered by 
                <a
                  href="https://github.com/auth-os/beta"
                >
                  Auth-os
                </a>
                .
              </p>
            </div>
            <div
              className="hm-Home_MainInfoButtonContainer"
            >
              <button
                className="hm-Home_BtnNew"
                onClick={[Function]}
              >
                New crowdsale
              </button>
              <div
                className="hm-Home_BtnChoose"
                onClick={[Function]}
              >
                Choose Crowdsale
              </div>
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>
  <div
    className="css-nm2dr3"
  >
    <span />
  </div>
</div>
`;

exports[`App Index should render screen with render without throwing an error 1`] = `"<div><section class=\\"hm-Home\\"><div class=\\"hm-Home_Scroll\\"><div class=\\"hm-Home_Top\\"><a class=\\"sw-LogoPrimary\\" href=\\"/\\"></a></div><div class=\\"hm-Home_MainInfoContainer\\"><div class=\\"hm-Home_MainInfo\\"><div class=\\"hm-Home_MainInfoTextContainer\\"><h1 class=\\"hm-Home_MainInfoTitle\\">Welcome to Token Wizard</h1><p class=\\"hm-Home_MainInfoDescription\\">Token Wizard is a client side tool to create ERC20 token and crowdsale in five steps. It helps you to publish contracts on the Ethereum network, create a crowdsale page with stats. For participants, the wizard creates a page to contribute into the campaign.</p><p class=\\"hm-Home_MainInfoPoweredBy\\">Token Wizard is powered by <a href=\\"https://github.com/auth-os/beta\\">Auth-os</a>.</p></div><div class=\\"hm-Home_MainInfoButtonContainer\\"><button class=\\"hm-Home_BtnNew\\">New crowdsale</button><div class=\\"hm-Home_BtnChoose\\">Choose Crowdsale</div></div></div></div></div></section></div><div class=\\"css-nm2dr3\\"><span></span></div>"`;
